bugbug/models/duplicate.py

```diff
@@ -0,0 +1,135 @@
+# -*- coding: utf-8 -*-
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from itertools import islice
+
+from imblearn.over_sampling import BorderlineSMOTE
+from sklearn.calibration import CalibratedClassifierCV
+from sklearn.compose import ColumnTransformer
+from sklearn.pipeline import Pipeline
+from sklearn.preprocessing import LabelEncoder
+from sklearn.svm import LinearSVC
+
+from bugbug import bug_features, bugzilla, feature_cleanup
+from bugbug.model import BugCoupleModel
+
+
+class LinearSVCWithLabelEncoding(CalibratedClassifierCV):
+    def __init__(self, clf):
+        super().__init__(clf)
+        self._le = LabelEncoder()
+
+    def fit(self, X, y):
+        super().fit(X, y)
+        self._le.fit(y)
+
+
+class DuplicateModel(BugCoupleModel):
+    def __init__(self, lemmatization=False):
+        BugCoupleModel.__init__(self, lemmatization)
+
+        self.sampler = BorderlineSMOTE(random_state=0)
+        self.calculate_importance = False
+
+        cleanup_functions = [
+            feature_cleanup.responses(),
+            feature_cleanup.hex(),
+            feature_cleanup.dll(),
+            feature_cleanup.fileref(),
+            feature_cleanup.url(),
+            feature_cleanup.synonyms(),
+            feature_cleanup.crash(),
+        ]
+
+        self.extraction_pipeline = Pipeline(
+            [
+                ("bug_extractor", bug_features.BugExtractor([], cleanup_functions)),
+                (
+                    "union",
+                    ColumnTransformer([("text", self.text_vectorizer(), "text")]),
+                ),
+            ]
+        )
+
+        self.clf = LinearSVCWithLabelEncoding(LinearSVC())
+
+    def get_labels(self):
+        bugs = []
+
+        for bug_data in islice(bugzilla.get_bugs(), 0, 20000):
+            bugs.append(bug_data["id"])
+
+        classes = {}
+        # Only store id's of bugs that have duplicate or are duplicate
+        duplicate_ids = []
+
+        # Store all remaining ids
+        all_ids = []
+
+        count = 4500
+        for bug_data in bugzilla.get_bugs():
+
+            if bug_data["duplicates"] != []:
+                duplicate_ids.append(bug_data["id"])
+                for duplicate_bug in bug_data["duplicates"]:
+                    if duplicate_bug in bugs:
+                        duplicate_ids.append(duplicate_bug)
+                        classes[(bug_data["id"], duplicate_bug)] = 1
```

Comment:
Every bug in a group of duplicates is a duplicate of every other bug in the same group. Here, you are only setting the label to 1 for the main bug and all other bugs of the group, we could instead set the label to 1 for all possible combinations of the bugs in the group.

You can use itertools for generating combinations.

Reply:
Actually, since we are limiting the number to 4500, let's not do this or we risk overfitting. We need to file an issue about this though, so that we do it in the future.

Reply:
Filed https://github.com/mozilla/bugbug/issues/522 about this.